# name       : bug813587.patch
# maintainer : Alexey
#
# Fix for LP bug #813587 / MySQL bug #51196 / MySQL bug #61790
#
# Clear MySQL connection errors in ha_federated::close(), since they
# can affect queries on other tables due to table cache eviction.
#
diff -ruN Percona-Server.orig/storage/federated/ha_federated.cc Percona-Server/storage/federated/ha_federated.cc
--- Percona-Server.orig/storage/federated/ha_federated.cc	2011-05-11 15:54:12.000000000 +0400
+++ Percona-Server/storage/federated/ha_federated.cc	2011-07-21 13:07:30.000000000 +0400
@@ -1675,6 +1675,8 @@ int ha_federated::open(const char *name,
 
 int ha_federated::close(void)
 {
+  THD *thd= current_thd;
+
   DBUG_ENTER("ha_federated::close");
 
   free_result();
@@ -1685,6 +1687,10 @@ int ha_federated::close(void)
   mysql_close(mysql);
   mysql= NULL;
 
+  /* Clear possible errors from mysql_close(), see LP bug #813587. */
+  if (thd)
+    thd->clear_error();
+
   DBUG_RETURN(free_share(share));
 }
 
